src/poly/commitment.rs

```diff
@@ -21,6 +23,95 @@ pub struct OpeningProof<C: CurveAffine> {
     z2: C::Scalar,
 }
 
+/// A multiscalar multiplication in the polynomial commitment scheme
+#[derive(Debug)]
+pub struct MSM<C: CurveAffine> {
+    /// TODO: documentation
+    pub g_scalars: Option<Vec<C::Scalar>>,
+
+    /// TODO: documentation
+    pub h_scalar: Option<C::Scalar>,
+
+    /// TODO: documentation
+    pub other_scalars: Vec<C::Scalar>,
+
+    /// TODO: documentation
+    pub other_bases: Vec<C>,
+}
+
+impl<'a, C: CurveAffine> MSM<C> {
+    /// Empty MSM
+    pub fn default(params: &Params<C>) -> Self {
+        let g_scalars = Some(vec![C::Scalar::one(); params.n as usize]);
+        let h_scalar = Some(C::Scalar::one());
+        let other_scalars: Vec<C::Scalar> = Vec::with_capacity(params.k as usize * 2 + 3);
+        let other_bases: Vec<C> = Vec::with_capacity(params.k as usize * 2 + 3);
+
+        MSM {
+            g_scalars,
+            h_scalar,
+            other_scalars,
+            other_bases,
+        }
+    }
+
+    /// Add arbitrary term (the scalar and the point)
+    pub fn add_term(&mut self, scalar: C::Scalar, point: C) {
+        &self.other_scalars.push(scalar);
+        &self.other_bases.push(point);
+    }
+
+    /// Add a vector of scalars to `g_scalars`
+    pub fn add_to_g(&mut self, scalars: Vec<C::Scalar>) {
+        for (g_scalar, scalar) in self
+            .g_scalars
+            .as_mut()
+            .unwrap()
+            .iter_mut()
+            .zip(scalars.iter())
+        {
+            *g_scalar += &scalar;
+        }
+    }
+
+    /// Add term to h
+    pub fn add_to_h(&mut self, scalar: C::Scalar) {
+        self.h_scalar = Some(self.h_scalar.unwrap() + &scalar);
+    }
+
+    /// Scale all scalars in the MSM by a random blinding factor
+    pub fn scale(&mut self, factor: C::Scalar) {
+        for g_scalar in self.g_scalars.as_mut().unwrap().iter_mut() {
+            *g_scalar *= &factor;
+        }
+        for other_scalar in self.other_scalars.iter_mut() {
+            *other_scalar *= &factor;
+        }
+        self.h_scalar = Some(self.h_scalar.unwrap() * &factor);
+    }
+
+    /// Perform multiexp and check that it results in zero
+    pub fn is_zero(&self, params: &'a Params<C>) -> bool {
+        let mut scalars: Vec<C::Scalar> = vec![];
+        let mut bases: Vec<C> = vec![];
```

Comment:
nit: It's possible to do `Vec::with_capacity` on these since we know in advance how many scalars/bases we have.

```rust
let len = self.g_scalars.as_ref().map(|v| v.len()).unwrap_or(0)
          + self.h_scalar.map(|_| 1).unwrap_or(0)
          + self.other_scalars.len();
let mut scalars: Vec<C::Scalar> = Vec::with_capacity(len);
let mut bases: Vec<C> = Vec::with_capacity(len);
```